### PR TITLE
File association

### PR DESCRIPTION
**Pull request by agritheory**
[Video demo](https://share.descript.com/view/1672YJBf8g8)

Please squash and merge

### CODE REVIEW

File: cloud_storage/cloud_storage/overrides/file.py [hunk 1]
```diff
@@ -50,17 +52,74 @@ def on_trash(self) -> None:
...
 			)
-
+		if len(self.file_association) > 1:
+			frappe.throw(
+				_("This file is attached to multiple documents and cannot be deleted"),
+				frappe.PermissionError,
+			)
 		super().on_trash()
 
+	def associate_files(self) -> File:
```
Comment (Alchez) [inline cloud_storage/cloud_storage/overrides/file.py]: I don't think this is correct. The function doesn't seem to be returning anything.

File: cloud_storage/cloud_storage/overrides/file.py [hunk 2]
```diff
@@ -50,17 +52,74 @@ def on_trash(self) -> None:
...
+				"File",
+				{"content_hash": self.content_hash, "name": ["!=", self.name], "is_folder": False} # type: ignore
+			)
+
+		if associated_doc:
+			existing_file = frappe.get_doc("File", associated_doc)
+			existing_file.attached_to_doctype = self.attached_to_doctype
+			existing_file.attached_to_name = self.attached_to_name
+			self.content_hash = existing_file.content_hash
+			self = existing_file
```
Comment (Alchez) [inline cloud_storage/cloud_storage/overrides/file.py:83]: Why are we overriding the instance? This probably needs a comment.

File: cloud_storage/cloud_storage/overrides/file.py [hunk 3]
```diff
@@ -50,17 +52,74 @@ def on_trash(self) -> None:
...
+		if len(self.file_association) > 1:
+			frappe.throw(
+				_("This file is attached to multiple documents and cannot be deleted"),
+				frappe.PermissionError,
+			)
 		super().on_trash()
 
+	def associate_files(self) -> File:
+		if not self.content_hash and "/api/method/retrieve" in self.file_url:  # type: ignore
+			associated_doc = frappe.get_value("File", {"file_url": self.file_url}, "name")
```
Comment (Alchez) [inline cloud_storage/cloud_storage/overrides/file.py:70]: Wouldn't this return the current file?

File: cloud_storage/cloud_storage/overrides/file.py [hunk 4]
```diff
@@ -50,17 +52,74 @@ def on_trash(self) -> None:
...
+		else:
```
Comment (Alchez) [inline cloud_storage/cloud_storage/overrides/file.py:71]: So it'll enter this block on any one of the following:
- the file document has a content hash set
- the file URL is not the string the cloud storage API expects

Should both cases be handled this way?

File: cloud_storage/cloud_storage/overrides/file.py [hunk 5]
```diff
@@ -50,17 +52,74 @@ def on_trash(self) -> None:
...
+				lambda row: row.link_doctype == self.attached_to_doctype
+				and row.link_name == self.attached_to_name,
+				self.file_association,
+			)
+		)
+		if not existing_attachment:
+			self.append(
+				"file_association",
+				{"link_doctype": self.attached_to_doctype, "link_name": self.attached_to_name},
+			)
```
Comment (Alchez) [inline cloud_storage/cloud_storage/overrides/file.py:96]: If there is no `associated_doc`, this block will not do anything

File: cloud_storage/cloud_storage/overrides/file.py [hunk 6]
```diff
@@ -50,17 +52,74 @@ def on_trash(self) -> None:
...
 
+	def after_insert(self) -> File:
+		if self.attached_to_doctype and self.attached_to_name and not self.file_association:
+			if not self.content_hash and "/api/method/retrieve" in self.file_url:
+				associated_doc = frappe.get_value("File", {"file_url": self.file_url}, "name")
+			else:
+				associated_doc = frappe.db.get_value(
+					"File", {"content_hash": self.content_hash, "name": ["!=", self.name], "is_folder": False}
+				)
+			rename_doc(self.doctype, self.name, associated_doc, merge=True, force=True)
```
Comment (Alchez) [inline cloud_storage/cloud_storage/overrides/file.py:118]: Does the function handle the case where `associated_doc` is None?

File: cloud_storage/cloud_storage/overrides/file.py [hunk 7]
```diff
@@ -50,17 +52,74 @@ def on_trash(self) -> None:
...
+
+	def remove_file_association(self, dt: str, dn: str) -> None:
+		if len(self.file_association) <= 1:
+			self.delete()
+			return
+		for row in self.file_association:
+			if row.link_doctype == dt and row.link_name == dn:
+				frappe.delete_doc("File Association", row.name)
+				del row
```
Comment (Alchez) [inline cloud_storage/cloud_storage/overrides/file.py:130]: I think you can just do:

```suggestion
				self.remove(row)
```

**Review by Alchez — CHANGES_REQUESTED** [review]

File: cloud_storage/cloud_storage/overrides/file.py (hunk 1 above)
Comment (agritheory) [reply cloud_storage/cloud_storage/overrides/file.py]: It used to, but it didn't have the desired effect and I didn't remove it.

File: cloud_storage/cloud_storage/overrides/file.py (hunk 3 above)
Comment (agritheory) [reply cloud_storage/cloud_storage/overrides/file.py:70]: No, the current file doesn't yet have a content hash. This only occurs when associating a second File (doctype) with the file (attachment).

File: cloud_storage/cloud_storage/overrides/file.py (hunk 4 above)
Comment (agritheory) [reply cloud_storage/cloud_storage/overrides/file.py:71]: Yes:
The first case covers the second upload of a file (attachment, jpeg, etc)
The second case covers when a File (doctype) is selected from the library.

File: cloud_storage/cloud_storage/overrides/file.py (hunk 5 above)
Comment (agritheory) [reply cloud_storage/cloud_storage/overrides/file.py:96]: Are you asking for a guard clause here?

File: cloud_storage/cloud_storage/overrides/file.py (hunk 6 above)
Comment (agritheory) [reply cloud_storage/cloud_storage/overrides/file.py:118]: It seems like you're asking for the same guard clause to solve this also

File: cloud_storage/cloud_storage/overrides/file.py (hunk 7 above)
Comment (agritheory) [reply cloud_storage/cloud_storage/overrides/file.py:130]: The docs say this should work, but it doesn't (and by docs, I mean that one time that Faris replied to my question about this on the forum, three or four years ago). `del` is more reliable in that you don't need the list's permission and matching function to remove it. Also more performant.

File: cloud_storage/cloud_storage/overrides/file.py (hunk 5 above)
Comment (agritheory) [reply cloud_storage/cloud_storage/overrides/file.py:96]: Re-read this and it should stay as-is; this prevents a second association of a file (attachment) with the File (doctype)

**Pull request merged**